File: backends/go-gin/internal/handler/client_handler.go
package handler

import (
	domclient "authorization-go/internal/domain/client"
	"authorization-go/internal/infrastructure/mail"
	uclient "authorization-go/internal/usecase/client"
	unotification "authorization-go/internal/usecase/notification"
	"authorization-go/pkg/apperror"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// ClientHandler はクライアント関連のHTTPハンドラーを提供します。
type ClientHandler struct {
	db          *gorm.DB
	newClientUC func(*gorm.DB) *uclient.Interactor
	newNotifUC  func(*gorm.DB) *unotification.Interactor
	mailer      *mail.Mailer
	historyRepo domclient.JwtHistoryRepository
	frontendURL string
}

// NewClientHandler は ClientHandler を生成します。
//
// db: GORM DB インスタンス
// newClientUC: クライアントユースケースファクトリ
// newNotifUC: 通知ユースケースファクトリ
// mailer: メール送信サービス
// historyRepo: JWT 履歴リポジトリ
// frontendURL: フロントエンド URL
func NewClientHandler(
	db *gorm.DB,
	newClientUC func(*gorm.DB) *uclient.Interactor,
	newNotifUC func(*gorm.DB) *unotification.Interactor,
	mailer *mail.Mailer,
	historyRepo domclient.JwtHistoryRepository,
	frontendURL string,
) *ClientHandler {
	return &ClientHandler{
		db:          db,
		newClientUC: newClientUC,
		newNotifUC:  newNotifUC,
		mailer:      mailer,
		historyRepo: historyRepo,
		frontendURL: frontendURL,
	}
}

// Index は検索条件に合致するクライアント一覧を返します。
// GET /api/clients
func (h *ClientHandler) Index(c *gin.Context) {
	cond := domclient.Condition{}

	if kw := c.Query("keyword"); kw != "" {
		cond.Keyword = &kw
	}
	if v := c.Query("start_from"); v != "" {
		if t, err := time.Parse("2006-01-02", v); err == nil {
			cond.StartFrom = &t
		}
	}
	if v := c.Query("start_to"); v != "" {
		if t, err := time.Parse("2006-01-02", v); err == nil {
			cond.StartTo = &t
		}
	}

	limit := 20
	if v := c.Query("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			limit = n
		}
	}
	page := 1
	if v := c.Query("page"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			page = n
		}
	}
	offset := limit * (page - 1)
	cond.Offset = offset
	cond.Limit = limit
	cond.Sort = c.Query("sort")
	cond.SortType = c.Query("sort_type")

	uc := h.newClientUC(h.db)
	count, err := uc.CountByCondition(cond)
	if err != nil {
		_ = c.Error(err)
		return
	}
	clients, err := uc.FindByCondition(cond)
	if err != nil {
		_ = c.Error(err)
		return
	}

	pager := BuildPager(count, limit, offset, len(clients))
	c.JSON(http.StatusOK, gin.H{
		"data":  mapClientList(clients),
		"pager": pager,
	})
}

// Show はIDでクライアント詳細を返します。
// GET /api/clients/:id
func (h *ClientHandler) Show(c *gin.Context) {
	id, err := parseUint64Param(c, "id")
	if err != nil {
		_ = c.Error(apperror.BadRequest("invalid_id"))
		return
	}
	client, err := h.newClientUC(h.db).FindByID(id)
	if err != nil {
		_ = c.Error(err)
		return
	}
	c.JSON(http.StatusOK, mapClientDetail(client))
}

// Store はクライアントを新規登録します。
// POST /api/clients/store
func (h *ClientHandler) Store(c *gin.Context) {
	var body StoreClientBody
	if err := c.ShouldBindJSON(&body); err != nil {
		_ = c.Error(apperror.UnprocessableEntity("validation_error"))
		return
	}
	if err := validateStruct(&body); err != nil {
		_ = c.Error(err)
		return
	}

	executorID := staffIDFromCookie(c)

	var storeVo *domclient.StoreVo
	if txErr := h.db.Transaction(func(tx *gorm.DB) error {
		var e error
		storeVo, e = h.newClientUC(tx).Store(uclient.StoreDto{
			Name:       body.Name,
			PostCode:   body.PostCode,
			Pref:       body.Pref,
			City:       body.City,
			Address:    body.Address,
			Building:   body.Building,
			Tel:        body.Tel,
			Email:      body.Email,
			ExecutorID: executorID,
		})
		return e
	}); txErr != nil {
		_ = c.Error(txErr)
		return
	}

	// 全スタッフへ通知配信（トランザクション外・ベストエフォート）
	notifURL := fmt.Sprintf("/clients/show?id=%d", storeVo.ID)
	_ = h.newNotifUC(h.db).FanOut(unotification.FanOutDto{
		Title:       "新しいクライアントが登録されました",
		Message:     storeVo.Name,
		MessageType: 1,
		ExecutorID:  executorID,
		URL:         notifURL,
	})

	// ご利用開始のご案内メール送信（非同期）
	activateURL := h.frontendURL + "/clients/" + storeVo.Identifier + "/qr"
	go h.mailer.SendActivation(storeVo.Email, storeVo.Name, activateURL)

	c.JSON(http.StatusCreated, gin.H{"id": storeVo.ID})
}

// Update はクライアントを更新して詳細を返します。
// PUT /api/clients/:id/update
func (h *ClientHandler) Update(c *gin.Context) {
	id, err := parseUint64Param(c, "id")
	if err != nil {
		_ = c.Error(apperror.BadRequest("invalid_id"))
		return
	}

	var body UpdateClientBody
	if err = c.ShouldBindJSON(&body); err != nil {
		_ = c.Error(apperror.UnprocessableEntity("validation_error"))
		return
	}
	if err = validateStruct(&body); err != nil {
		_ = c.Error(err)
		return
	}

	executorID := staffIDFromCookie(c)

	var detailVo *domclient.DetailVo
	if txErr := h.db.Transaction(func(tx *gorm.DB) error {
		var e error
		detailVo, e = h.newClientUC(tx).Update(uclient.UpdateDto{
			ID:         id,
			Name:       body.Name,
			PostCode:   body.PostCode,
			Pref:       body.Pref,
			City:       body.City,
			Address:    body.Address,
			Building:   body.Building,
			Tel:        body.Tel,
			Email:      body.Email,
			Status:     body.Status,
			ExecutorID: executorID,
		})
		return e
	}); txErr != nil {
		_ = c.Error(txErr)
		return
	}
	c.JSON(http.StatusOK, mapClientDetail(detailVo))
}

// Qr はidentifierでQRコードデータを返します。
// GET /api/clients/:id/qr
func (h *ClientHandler) Qr(c *gin.Context) {
	identifier := c.Param("id")
	vo, err := h.newClientUC(h.db).GetQr(identifier)
	if err != nil {
		_ = c.Error(err)
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"identifier":   vo.Identifier,
		"deeplink_url": vo.DeeplinkURL,
	})
}

// Info はidentifierでスマホアプリ向けクライアント情報を返します。
// GET /api/clients/:id/info
func (h *ClientHandler) Info(c *gin.Context) {
	identifier := c.Param("id")
	vo, err := h.newClientUC(h.db).GetInfo(identifier)
	if err != nil {
		_ = c.Error(err)
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"identifier": vo.Identifier,
		"name":       vo.Name,
		"status":     vo.Status,
	})
}

// Start は利用開始処理を行い、アクセストークンを返します。
// PATCH /api/clients/:id/start
func (h *ClientHandler) Start(c *gin.Context) {
	identifier := c.Param("id")
	var startVo *domclient.StartVo
	if txErr := h.db.Transaction(func(tx *gorm.DB) error {
		var e error
		startVo, e = h.newClientUC(tx).Start(identifier)
		return e
	}); txErr != nil {
		_ = c.Error(txErr)
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"access_token": startVo.AccessToken,
	})
}

// Stop は利用停止処理を行います。
// PATCH /api/clients/:id/stop
func (h *ClientHandler) Stop(c *gin.Context) {
	identifier := c.Param("id")
	if txErr := h.db.Transaction(func(tx *gorm.DB) error {
		return h.newClientUC(tx).Stop(identifier)
	}); txErr != nil {
		_ = c.Error(txErr)
		return
	}
	c.JSON(http.StatusOK, gin.H{})
}

// Destroy はクライアントを論理削除します。
// DELETE /api/clients/:id/delete
func (h *ClientHandler) Destroy(c *gin.Context) {
	id, err := parseUint64Param(c, "id")
	if err != nil {
		_ = c.Error(apperror.BadRequest("invalid_id"))
		return
	}
	executorID := staffIDFromCookie(c)
	if txErr := h.db.Transaction(func(tx *gorm.DB) error {
		return h.newClientUC(tx).Destroy(id, executorID)
	}); txErr != nil {
		_ = c.Error(txErr)
		return
	}
	c.JSON(http.StatusOK, gin.H{})
}

// JwtHistories はクライアントの JWT 履歴一覧を返します。
// GET /api/clients/:id/jwt-histories
func (h *ClientHandler) JwtHistories(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_id"})
		return
	}
	histories, err := h.historyRepo.FindByClientID(id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "internal_error"})
		return
	}
	out := make([]gin.H, 0, len(histories))
	for _, hist := range histories {
		out = append(out, gin.H{
			"id":        hist.ID,
			"member_id": hist.MemberID,
			"issue_at":  hist.IssueAt.Format("2006-01-02 15:04:05"),
			"jwt":       hist.Jwt,
		})
	}
	c.JSON(http.StatusOK, out)
}

// ---------- 変換ヘルパー ----------

// mapClientList はクライアント一覧 Vo をレスポンス用マップのスライスに変換します。
func mapClientList(clients []*domclient.ListItem) []gin.H {
	out := make([]gin.H, 0, len(clients))
	for _, c := range clients {
		out = append(out, gin.H{
			"id":         c.ID,
			"name":       c.Name,
			"status":     c.Status,
			"start_at":   formatTimePtr(c.StartAt),
			"stop_at":    formatTimePtr(c.StopAt),
			"created_at": formatTime(c.CreatedAt),
			"updated_at": formatTime(c.UpdatedAt),
		})
	}
	return out
}

// mapClientDetail はクライアント詳細 Vo をレスポンス用マップに変換します。
func mapClientDetail(c *domclient.DetailVo) gin.H {
	return gin.H{
		"id":         c.ID,
		"name":       c.Name,
		"identifier": c.Identifier,
		"post_code":  c.PostCode,
		"pref":       c.Pref,
		"city":       c.City,
		"address":    c.Address,
		"building":   c.Building,
		"tel":        c.Tel,
		"email":      c.Email,
		"status":     c.Status,
		"start_at":   formatTimePtr(c.StartAt),
		"stop_at":    formatTimePtr(c.StopAt),
		"created_at": formatTime(c.CreatedAt),
		"updated_at": formatTime(c.UpdatedAt),
	}
}

// parseUint64Param はパスパラメータを uint64 に変換します。
func parseUint64Param(c *gin.Context, key string) (uint64, error) {
	return strconv.ParseUint(c.Param(key), 10, 64)
}

File: backends/csharp-aspnet/src/Authorization.Api/Handler/ClientHandler.cs
/*
 * クライアントハンドラーモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
using System.Text.RegularExpressions;
using Authorization.Api.Config;
using Authorization.Api.Domain.Client;
using Authorization.Api.Domain.Notification;
using Authorization.Api.Infrastructure.Mail;
using Authorization.Api.Support;
using Authorization.Api.UseCase.Client;
using Authorization.Api.UseCase.Notification;
using static Authorization.Api.Handler.HttpHelpers;

namespace Authorization.Api.Handler;

/// <summary>クライアント登録/更新ボディの検証です。</summary>
public static partial class ClientValidation
{
    [GeneratedRegex(@"^\d{10,11}$")]
    private static partial Regex TelPattern();

    [GeneratedRegex(@"^[^\s@]+@[^\s@]+\.[^\s@]+$")]
    private static partial Regex EmailPattern();

    /// <summary>登録ボディを検証します。</summary>
    /// <param name="name">名称</param>
    /// <param name="postCode">郵便番号</param>
    /// <param name="pref">都道府県</param>
    /// <param name="city">市区町村</param>
    /// <param name="address">番地</param>
    /// <param name="building">建物名（任意）</param>
    /// <param name="tel">電話番号</param>
    /// <param name="email">メールアドレス</param>
    /// <returns>全項目が妥当な場合は true</returns>
    public static bool ValidateStore(string name, string postCode, string pref, string city, string address,
        string? building, string tel, string email) =>
        Required(name, 255) && Required(postCode, 8) && Required(pref, 50) && Required(city, 100) &&
        Required(address, 255) && Optional(building, 255) &&
        tel.Length > 0 && TelPattern().IsMatch(tel) &&
        email.Length > 0 && email.Length <= 255 && EmailPattern().IsMatch(email);

    /// <summary>更新ボディを検証します（存在する項目のみ）。</summary>
    /// <param name="name">名称（未指定なら検証しない）</param>
    /// <param name="postCode">郵便番号（未指定なら検証しない）</param>
    /// <param name="pref">都道府県（未指定なら検証しない）</param>
    /// <param name="city">市区町村（未指定なら検証しない）</param>
    /// <param name="address">番地（未指定なら検証しない）</param>
    /// <param name="building">建物名（未指定なら検証しない）</param>
    /// <param name="tel">電話番号（未指定なら検証しない）</param>
    /// <param name="email">メールアドレス（未指定なら検証しない）</param>
    /// <returns>指定された項目が全て妥当な場合は true</returns>
    public static bool ValidateUpdate(string? name, string? postCode, string? pref, string? city, string? address,
        string? building, string? tel, string? email) =>
        Optional(name, 255) && Optional(postCode, 8) && Optional(pref, 50) && Optional(city, 100) &&
        Optional(address, 255) && Optional(building, 255) &&
        (tel is null || TelPattern().IsMatch(tel)) &&
        (email is null || (email.Length <= 255 && EmailPattern().IsMatch(email)));

    private static bool Required(string v, int max) => v.Length > 0 && v.Length <= max;
    private static bool Optional(string? v, int max) => v is null || v.Length <= max;
}

/// <summary>クライアントハンドラーです。</summary>
public sealed class ClientHandler(
    ClientInteractor clientUC,
    NotificationInteractor notificationUC,
    IMailer mailer,
    IJwtHistoryRepository jwtHistoryRepo,
    AppSettings app)
{
    /// <summary>クライアント一覧を返します。</summary>
    /// <param name="req">HTTPリクエスト（keyword/start_from/start_to/statuses/limit/page/sort/sort_typeを使用）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>一覧データとページャー情報のJSON</returns>
    public async Task<IResult> IndexAsync(HttpRequest req, CancellationToken ct)
    {
        var limit = Math.Max(1, QueryInt(req, "limit") ?? 10);
        var page  = Math.Max(1, QueryInt(req, "page") ?? 1);
        if (SafeOffset(limit, page) is not int offset) return Error(400, "page_out_of_range");
        var statuses = req.Query["statuses"].Concat(req.Query["statuses[]"])
            .SelectMany(v => (v ?? "").Split(','))
            .Select(s => int.TryParse(s.Trim(), out var r) ? r : (int?)null)
            .OfType<int>()
            .ToList();

        var dto = new ClientListConditionDto(
            Keyword:   Query(req, "keyword"),
            StartFrom: Query(req, "start_from"),
            StartTo:   Query(req, "start_to"),
            Statuses:  statuses,
            Offset:    offset,
            Limit:     limit,
            Sort:      Query(req, "sort"),
            SortType:  Query(req, "sort_type"));

        var (items, count) = await clientUC.FindByConditionWithCountAsync(dto, ct);
        var data = items.Select(c => new Dictionary<string, object?>
        {
            ["id"]         = c.Id,
            ["name"]       = c.Name,
            ["status"]     = c.Status,
            ["start_at"]   = DateFormat.ToMinute(c.StartAt),
            ["stop_at"]    = DateFormat.ToMinute(c.StopAt),
            ["created_at"] = DateFormat.ToMinute(c.CreatedAt),
            ["updated_at"] = DateFormat.ToMinute(c.UpdatedAt),
        }).ToList();

        return Results.Json(new Dictionary<string, object?>
        {
            ["data"]  = data,
            ["pager"] = Pager.Build(count, limit, offset, items.Count).ToJson(),
        });
    }

    /// <summary>クライアント詳細を返します。</summary>
    /// <param name="id">クライアントID（文字列）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>クライアント詳細のJSON、IDが不正な場合は400</returns>
    public async Task<IResult> ShowAsync(string id, CancellationToken ct)
    {
        if (!long.TryParse(id, out var clientId)) return InvalidId();
        var c = await clientUC.FindByIdAsync(clientId, ct);
        return Results.Json(DetailJson(c));
    }

    /// <summary>
    /// クライアントを登録します。登録後、全スタッフへ通知を配信し、利用開始案内メールを送信します。
    /// </summary>
    /// <param name="req">HTTPリクエストボディ（name/post_code/pref/city/address/building/tel/email）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>登録したクライアントIDのJSON（201）、検証エラーの場合は422</returns>
    public async Task<IResult> StoreAsync(HttpRequest req, CancellationToken ct)
    {
        var executorId = StaffId(req);
        var body = await ReadJsonObjectAsync(req, ct);

        var name     = body.Str("name") ?? "";
        var postCode = body.Str("post_code") ?? "";
        var pref     = body.Str("pref") ?? "";
        var city     = body.Str("city") ?? "";
        var address  = body.Str("address") ?? "";
        var building = body.Str("building");
        var tel      = body.Str("tel") ?? "";
        var email    = body.Str("email") ?? "";

        if (!ClientValidation.ValidateStore(name, postCode, pref, city, address, building, tel, email))
        {
            return Error(422, "validation_error");
        }

        var client = await clientUC.StoreAsync(new ClientStoreDto(
            name, postCode, pref, city, address, building ?? "", tel, email, executorId), ct);

        await notificationUC.FanOutAsync(new NotificationFanOutDto(
            MessageType: NotificationMessageType.Info,
            Title:       "新しいクライアントが登録されました",
            Message:     client.Name,
            ExecutorId:  executorId,
            Url:         $"/clients/show?id={client.Id}"), ct);

        var activateUrl = $"{app.FrontendUrl}/clients/{client.Identifier}/qr";
        _ = mailer.SendActivationAsync(client.Email, client.Name, activateUrl, CancellationToken.None);

        return Results.Json(new Dictionary<string, object?> { ["id"] = client.Id }, statusCode: 201);
    }

    /// <summary>クライアントを更新します（楽観排他ロック）。</summary>
    /// <param name="id">クライアントID（文字列）</param>
    /// <param name="req">HTTPリクエストボディ（更新するフィールドとversion）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>更新後のクライアント詳細のJSON、IDが不正または検証エラーの場合は400/422</returns>
    /// <exception cref="AppException">存在しない場合（404）、バージョン不一致（409）</exception>
    public async Task<IResult> UpdateAsync(string id, HttpRequest req, CancellationToken ct)
    {
        if (!long.TryParse(id, out var clientId)) return InvalidId();
        var executorId = StaffId(req);
        var body = await ReadJsonObjectAsync(req, ct);

        var name     = body.Str("name");
        var postCode = body.Str("post_code");
        var pref     = body.Str("pref");
        var city     = body.Str("city");
        var address  = body.Str("address");
        var building = body.Str("building");
        var tel      = body.Str("tel");
        var email    = body.Str("email");
        var status   = body.Int("status");
        var version  = body.Int("version");

        if (!ClientValidation.ValidateUpdate(name, postCode, pref, city, address, building, tel, email))
        {
            return Error(422, "validation_error");
        }
        if (version is not int v) return Error(400, "version_required");

        var dto = new ClientUpdateDto(clientId, name, postCode, pref, city, address, building, tel, email, status,
            executorId, v);
        var c = await clientUC.UpdateAsync(dto, ct);
        return Results.Json(DetailJson(c));
    }

    /// <summary>QRコード用データを返します。</summary>
    /// <param name="identifier">クライアント識別子</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>識別子とディープリンクURLのJSON</returns>
    /// <exception cref="AppException">存在しない場合（404）</exception>
    public async Task<IResult> QrAsync(string identifier, CancellationToken ct)
    {
        var vo = await clientUC.GetQrAsync(new ClientQrDto(identifier), ct);
        return Results.Json(new Dictionary<string, object?>
        {
            ["identifier"]   = vo.Identifier,
            ["deeplink_url"] = vo.DeeplinkUrl,
        });
    }

    /// <summary>スマホアプリ向けクライアント情報を返します。</summary>
    /// <param name="identifier">クライアント識別子</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>識別子・名称・状態のJSON</returns>
    /// <exception cref="AppException">存在しない場合（404）</exception>
    public async Task<IResult> InfoAsync(string identifier, CancellationToken ct)
    {
        var vo = await clientUC.GetInfoAsync(new ClientInfoDto(identifier), ct);
        return Results.Json(new Dictionary<string, object?>
        {
            ["identifier"] = vo.Identifier,
            ["name"]       = vo.Name,
            ["status"]     = vo.Status,
        });
    }

    /// <summary>利用開始し、アクセストークンを返します。</summary>
    /// <param name="identifier">クライアント識別子</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>アクセストークンのJSON</returns>
    /// <exception cref="AppException">存在しない場合（404）</exception>
    public async Task<IResult> StartAsync(string identifier, CancellationToken ct)
    {
        var vo = await clientUC.StartAsync(new ClientStartDto(identifier), ct);
        return Results.Json(new Dictionary<string, object?> { ["access_token"] = vo.AccessToken });
    }

    /// <summary>利用停止します。</summary>
    /// <param name="identifier">クライアント識別子</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>空レスポンス</returns>
    /// <exception cref="AppException">存在しない場合（404）</exception>
    public async Task<IResult> StopAsync(string identifier, CancellationToken ct)
    {
        await clientUC.StopAsync(new ClientStopDto(identifier), ct);
        return Empty();
    }

    /// <summary>クライアントを論理削除します。</summary>
    /// <param name="id">クライアントID（文字列）</param>
    /// <param name="req">HTTPリクエストボディ（version）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>空レスポンス、IDが不正な場合は400</returns>
    /// <exception cref="AppException">バージョン未指定（400）、存在しない場合（404）、バージョン不一致（409）</exception>
    public async Task<IResult> DestroyAsync(string id, HttpRequest req, CancellationToken ct)
    {
        if (!long.TryParse(id, out var clientId)) return InvalidId();
        var body = await ReadJsonObjectAsync(req, ct);
        await clientUC.DestroyAsync(clientId, StaffId(req), body.Int("version"), ct);
        return Empty();
    }

    /// <summary>クライアントのJWT発行履歴一覧を返します。</summary>
    /// <param name="id">クライアントID（文字列）</param>
    /// <param name="req">HTTPリクエスト（limit/page/sort/sort_typeを使用）</param>
    /// <param name="ct">キャンセレーショントークン</param>
    /// <returns>履歴一覧とページャー情報のJSON、IDが不正な場合は400</returns>
    public async Task<IResult> JwtHistoriesAsync(string id, HttpRequest req, CancellationToken ct)
    {
        if (!long.TryParse(id, out var clientId)) return InvalidId();
        var limit = Math.Max(1, QueryInt(req, "limit") ?? 10);
        var page  = Math.Max(1, QueryInt(req, "page") ?? 1);
        if (SafeOffset(limit, page) is not int offset) return Error(400, "page_out_of_range");

        var cond = new JwtHistoryCondition(clientId, offset, limit,
            Query(req, "sort") ?? "issue_at", Query(req, "sort_type") ?? "desc");

        var count     = await jwtHistoryRepo.CountByConditionAsync(cond, ct);
        var histories = await jwtHistoryRepo.FindByConditionAsync(cond, ct);
        var data = histories.Select(h => new Dictionary<string, object?>
        {
            ["id"]        = h.Id,
            ["member_id"] = h.MemberId,
            ["issue_at"]  = DateFormat.ToSecond(h.IssueAt),
            ["jwt"]       = h.Jwt,
        }).ToList();

        return Results.Json(new Dictionary<string, object?>
        {
            ["data"]  = data,
            ["pager"] = Pager.Build(count, limit, offset, histories.Count).ToJson(),
        });
    }

    /// <summary>クライアント詳細を JSON 用の辞書に変換します。</summary>
    /// <param name="c">クライアント詳細</param>
    /// <returns>JSON化用の辞書</returns>
    private static Dictionary<string, object?> DetailJson(ClientDetailVo c) => new()
    {
        ["id"]         = c.Id,
        ["name"]       = c.Name,
        ["identifier"] = c.Identifier,
        ["post_code"]  = c.PostCode,
        ["pref"]       = c.Pref,
        ["city"]       = c.City,
        ["address"]    = c.Address,
        ["building"]   = c.Building,
        ["tel"]        = c.Tel,
        ["email"]      = c.Email,
        ["status"]     = c.Status,
        ["start_at"]   = DateFormat.ToMinute(c.StartAt),
        ["stop_at"]    = DateFormat.ToMinute(c.StopAt),
        ["created_at"] = DateFormat.ToMinute(c.CreatedAt),
        ["updated_at"] = DateFormat.ToMinute(c.UpdatedAt),
        ["version"]    = c.Version,
    };
}
